extensions/amp-analytics/0.1/visibility-impl.js

```diff
@@ -508,29 +545,55 @@ export class Visibility {
       this.setState_(state, visible, 0);
     }
 
-    const waitForContinuousTime = config[CONTINUOUS_TIME_MIN]
-        ? config[CONTINUOUS_TIME_MIN] - state[CONTINUOUS_TIME]
-        : Infinity;
-    const waitForTotalTime = config[TOTAL_TIME_MIN]
-        ? config[TOTAL_TIME_MIN] - state[TOTAL_VISIBLE_TIME]
-        : Infinity;
+    listener['state'] = state;
+
+    return this.isConditionMet_(state, config, triggerType);
+  }
+
+  /**
+   * @param {!Object} state
+   * @param {!Object} config
+   * @return {number}
+   * @private
+   */
+  computeTimeToWait_(state, config) {
+    const waitForContinuousTime =
+        config[CONTINUOUS_TIME_MIN] > state[CONTINUOUS_TIME]
+            ? config[CONTINUOUS_TIME_MIN] - state[CONTINUOUS_TIME]
+            : null;
+
+    const waitForTotalTime =
+        config[TOTAL_TIME_MIN] > state[TOTAL_VISIBLE_TIME]
+            ? config[TOTAL_TIME_MIN] - state[TOTAL_VISIBLE_TIME]
+            : null;
 
     // Wait for minimum of (previous timeToWait, positive values of
     // waitForContinuousTime and waitForTotalTime).
     this.timeToWait_ = Math.min(this.timeToWait_,
-        waitForContinuousTime > 0 ? waitForContinuousTime : Infinity,
-        waitForTotalTime > 0 ? waitForTotalTime : Infinity);
-    listener['state'] = state;
+        waitForContinuousTime || Infinity,
```

Comment:
This is the V1 logic that I didn't change. My understanding of the logic is: it repeats the work every `this.timeToWait_` milliseconds, until the trigger fires. This value can only be reduced, due to the change on `waitForContinuousTime` and `waitForTotalTime`.

The algorithm runs much more loops than it should, though it gives the same triggering results, but definitely can be improved. 

This issue has been addressed in the V2 implementation.